### ISSUE TITLE
improve logic to determine if a mirror is "healthy"

### ISSUE THREAD

**Issue by sobomax** [lagged/apt-spy2#78]
Hi,

We are using apt-spy2 in our GitHub actions to sanitize list of ubuntu mirrors before running our test code. However, since yesterday it inserts what seems to be dis-functional mirror into the list, which is then fails when running apt-get update. It seems like liveness check algorithm is inadequate to figure out which mirrors are good, defeating the whole purpose of using apt-spy2. :(

```
+ sudo apt-spy2 check
Mirror: http://mirror.funkfreundelandshut.de/ubuntu/ - UP
Mirror: http://ftp.stw-bonn.de/ubuntu/ - UP
Mirror: http://mirror.eu-fr.kamatera.com/ubuntu/ - UP
Mirror: https://debian.charite.de/ubuntu/ - UP
Mirror: http://mirror.daniel-jost.net/ubuntu/ - UP
Mirror: http://ftp.uni-kl.de/pub/linux/ubuntu/ - UP
Mirror: http://mirror.ipb.de/ubuntu/ - UP
Mirror: https://ftp.uni-stuttgart.de/ubuntu/ - UP
Mirror: https://mirror.creoline.net/ubuntu/ - UP
Mirror: http://ftp.hosteurope.de/mirror/archive.ubuntu.com/ - UP
Mirror: http://mirror.wtnet.de/ubuntu/ - UP
Mirror: http://artfiles.org/ubuntu.com/ - UP
Mirror: http://mirror.kamp.de/ubuntu/ - UP
Mirror: http://ftp.rz.tu-bs.de/pub/mirror/ubuntu-packages/ - UP
Mirror: https://mirror.cxserv.de/ubuntu/ - UP
Mirror: https://mirror.scaleuptech.com/ubuntu/ - UP
Mirror: http://ftp.tu-ilmenau.de/mirror/ubuntu/ - UP
Mirror: https://mirror.netcologne.de/ubuntu/ - UP
Mirror: https://mirrors.xtom.de/ubuntu/ - UP
Mirror: https://mirror.dogado.de/ubuntu/ - UP
Mirror: http://ftp.tu-chemnitz.de/pub/linux/ubuntu/ - UP
Mirror: https://mirror.kkg.berlin/ubuntu/ - UP
Mirror: https://mirror.netzwerge.de/ubuntu/ - UP
Mirror: http://ftp.fau.de/ubuntu/ - UP
Mirror: https://mirror.united-gameserver.de/ubuntu/ - UP
Mirror: https://de.mirrors.clouvider.net/ubuntu/ - UP
Mirror: https://mirror.de.vangus.co.il/ubuntu/ - DOWN
Mirror: http://ftp.uni-mainz.de/ubuntu/ - UP
Mirror: http://ftp.halifax.rwth-aachen.de/ubuntu/ - UP
Mirror: http://linux.darkpenguin.net/distros/ubuntu-archive/ - UP
Mirror: https://ubuntu.brontocdn.com/ - UP
Mirror: http://packages.oth-regensburg.de/ubuntu/ - UP
Mirror: http://ftp.uni-bayreuth.de/linux/ubuntu/ubuntu/ - UP
Mirror: http://suse.uni-leipzig.de/pub/releases.ubuntu.com/ubuntu/ - UP
Mirror: http://mirror.23media.com/ubuntu/ - UP
Mirror: http://ubuntu.mirror.lrz.de/ubuntu/ - UP
Mirror: http://ftp5.gwdg.de/pub/linux/debian/ubuntu/ - UP
Mirror: http://ubuntu.mirror.tudos.de/ubuntu/ - UP
Mirror: http://mirror2.tuxinator.org/ubuntu/ - UP
Mirror: https://mirror.23m.com/ubuntu/ - UP
Mirror: https://mirror.de.leaseweb.net/ubuntu/ - UP
Mirror: http://ftp-stud.hs-esslingen.de/ubuntu/ - UP
Mirror: http://mirror.serverloft.eu/ubuntu/ubuntu/ - UP
Mirror: http://ftp.tu-chemnitz.de/pub/linux/ubuntu-ports/ - UP
Mirror: http://archive.ubuntu.com/ubuntu/ - UP
+ sudo apt-spy2 fix --commit
The closest mirror is: http://mirror.funkfreundelandshut.de/ubuntu/
Updating /etc/apt/sources.list
Updated '/etc/apt/sources.list' with http://mirror.funkfreundelandshut.de/ubuntu/
Run `apt-get update` to update
+ sudo -H DEBIAN_FRONTEND=noninteractive apt-get update
Get:1 https://packages.microsoft.com/ubuntu/20.04/prod focal InRelease [10.5 kB]
Get:2 https://packages.microsoft.com/ubuntu/20.04/prod focal/main arm64 Packages [48.7 kB]
Get:3 https://packages.microsoft.com/ubuntu/20.04/prod focal/main amd64 Packages [[22](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:23)6 kB]
Hit:4 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu focal InRelease
Get:5 https://mirror.funkfreundelandshut.de/ubuntu focal InRelease [265 kB]
Get:6 https://mirror.funkfreundelandshut.de/ubuntu focal-updates InRelease [114 kB]
Get:7 https://mirror.funkfreundelandshut.de/ubuntu focal-backports InRelease [108 kB]
Get:8 https://mirror.funkfreundelandshut.de/ubuntu focal-security InRelease [114 kB]
Ign:9 https://mirror.funkfreundelandshut.de/ubuntu focal/main amd64 Packages
Ign:10 https://mirror.funkfreundelandshut.de/ubuntu focal/main Translation-en
Ign:9 https://mirror.funkfreundelandshut.de/ubuntu focal/main amd64 Packages
Ign:11 https://mirror.funkfreundelandshut.de/ubuntu focal/main amd64 c-n-f Metadata
Ign:10 https://mirror.funkfreundelandshut.de/ubuntu focal/main Translation-en
Ign:12 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted amd64 Packages
Ign:11 https://mirror.funkfreundelandshut.de/ubuntu focal/main amd64 c-n-f Metadata
Ign:13 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted Translation-en
Ign:12 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted amd64 Packages
Ign:14 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted amd64 c-n-f Metadata
Ign:13 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted Translation-en
Ign:15 https://mirror.funkfreundelandshut.de/ubuntu focal/universe amd64 Packages
Ign:14 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted amd64 c-n-f Metadata
Ign:16 https://mirror.funkfreundelandshut.de/ubuntu focal/universe Translation-en
Ign:15 https://mirror.funkfreundelandshut.de/ubuntu focal/universe amd64 Packages
Ign:17 https://mirror.funkfreundelandshut.de/ubuntu focal/universe amd64 c-n-f Metadata
Ign:16 https://mirror.funkfreundelandshut.de/ubuntu focal/universe Translation-en
Ign:18 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse amd64 Packages
Ign:17 https://mirror.funkfreundelandshut.de/ubuntu focal/universe amd64 c-n-f Metadata
Ign:19 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse Translation-en
Ign:18 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse amd64 Packages
Ign:19 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse Translation-en
Ign:20 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse amd64 c-n-f Metadata
Ign:20 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse amd64 c-n-f Metadata
Ign:21 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main amd64 Packages
Ign:21 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main amd64 Packages
Ign:22 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main Translation-en
Ign:22 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main Translation-en
Ign:[23](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:24) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main amd64 c-n-f Metadata
Ign:[24](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:25) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted amd64 Packages
Ign:[25](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:26) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted Translation-en
Ign:23 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main amd64 c-n-f Metadata
Ign:[26](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:27) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted amd64 c-n-f Metadata
Ign:24 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted amd64 Packages
Ign:[27](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:28) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe amd64 Packages
Ign:25 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted Translation-en
Ign:[28](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:29) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe Translation-en
Ign:26 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted amd64 c-n-f Metadata
Ign:[29](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:30) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe amd64 c-n-f Metadata
Ign:27 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe amd64 Packages
Ign:[30](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:31) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse amd64 Packages
Ign:28 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe Translation-en
Ign:[31](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:32) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse Translation-en
Ign:29 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe amd64 c-n-f Metadata
Ign:[32](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:33) https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse amd64 c-n-f Metadata
Ign:30 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse amd64 Packages
Ign:[33](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:34) https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main amd64 Packages
Ign:31 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse Translation-en
Ign:32 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse amd64 c-n-f Metadata
Ign:33 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main amd64 Packages
Ign:[34](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:35) https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main Translation-en
Ign:34 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main Translation-en
Ign:[35](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:36) https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main amd64 c-n-f Metadata
Ign:35 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main amd64 c-n-f Metadata
Ign:[36](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:37) https://mirror.funkfreundelandshut.de/ubuntu focal-backports/restricted amd64 c-n-f Metadata
Ign:36 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/restricted amd64 c-n-f Metadata
Ign:[37](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:38) https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe amd64 Packages
Ign:[38](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:39) https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe Translation-en
Ign:[39](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:40) https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe amd64 c-n-f Metadata
Ign:37 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe amd64 Packages
Ign:[40](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:41) https://mirror.funkfreundelandshut.de/ubuntu focal-backports/multiverse amd64 c-n-f Metadata
Ign:38 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe Translation-en
Ign:[41](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:42) https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 Packages
Ign:39 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe amd64 c-n-f Metadata
Ign:[42](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:43) https://mirror.funkfreundelandshut.de/ubuntu focal-security/main Translation-en
Ign:40 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/multiverse amd64 c-n-f Metadata
Ign:[43](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:44) https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 c-n-f Metadata
Ign:41 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 Packages
Ign:[44](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:45) https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted amd64 Packages
Ign:42 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main Translation-en
Ign:[45](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:46) https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted Translation-en
Ign:43 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 c-n-f Metadata
Ign:[46](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:47) https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted amd64 c-n-f Metadata
Ign:44 https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted amd64 Packages
Ign:[47](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:48) https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe amd64 Packages
Ign:45 https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted Translation-en
Ign:46 https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted amd64 c-n-f Metadata
Ign:47 https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe amd64 Packages
Ign:[48](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:49) https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe Translation-en
Ign:48 https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe Translation-en
Ign:[49](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:50) https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe amd64 c-n-f Metadata
Ign:49 https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe amd64 c-n-f Metadata
Ign:[50](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:51) https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse amd64 Packages
Ign:50 https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse amd64 Packages
Ign:[51](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:52) https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse Translation-en
Ign:[52](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:53) https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse amd64 c-n-f Metadata
Ign:9 https://mirror.funkfreundelandshut.de/ubuntu focal/main amd64 Packages
Ign:51 https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse Translation-en
Ign:10 https://mirror.funkfreundelandshut.de/ubuntu focal/main Translation-en
Ign:52 https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse amd64 c-n-f Metadata
Ign:11 https://mirror.funkfreundelandshut.de/ubuntu focal/main amd64 c-n-f Metadata
Ign:9 https://mirror.funkfreundelandshut.de/ubuntu focal/main amd64 Packages
Ign:12 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted amd64 Packages
Ign:10 https://mirror.funkfreundelandshut.de/ubuntu focal/main Translation-en
Ign:13 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted Translation-en
Err:11 https://mirror.funkfreundelandshut.de/ubuntu focal/main amd64 c-n-f Metadata
  404  Not Found [IP: 188.174.173.78 443]
Ign:14 https://mirror.funkfreundelandshut.de/ubuntu focal/restricted amd64 c-n-f Metadata
Ign:15 https://mirror.funkfreundelandshut.de/ubuntu focal/universe amd64 Packages
Ign:16 https://mirror.funkfreundelandshut.de/ubuntu focal/universe Translation-en
Ign:17 https://mirror.funkfreundelandshut.de/ubuntu focal/universe amd64 c-n-f Metadata
Ign:18 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse amd64 Packages
Ign:19 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse Translation-en
Ign:20 https://mirror.funkfreundelandshut.de/ubuntu focal/multiverse amd64 c-n-f Metadata
Ign:21 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main amd64 Packages
Ign:21 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main amd64 Packages
Ign:22 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main Translation-en
Ign:22 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main Translation-en
Ign:23 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main amd64 c-n-f Metadata
Err:23 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/main amd64 c-n-f Metadata
  404  Not Found [IP: 188.174.173.78 443]
Ign:24 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted amd64 Packages
Ign:25 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted Translation-en
Ign:26 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/restricted amd64 c-n-f Metadata
Ign:27 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe amd64 Packages
Ign:28 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe Translation-en
Ign:29 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/universe amd64 c-n-f Metadata
Ign:30 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse amd64 Packages
Ign:31 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse Translation-en
Ign:32 https://mirror.funkfreundelandshut.de/ubuntu focal-updates/multiverse amd64 c-n-f Metadata
Ign:33 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main amd64 Packages
Ign:34 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main Translation-en
Ign:33 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main amd64 Packages
Ign:35 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main amd64 c-n-f Metadata
Ign:34 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main Translation-en
Ign:36 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/restricted amd64 c-n-f Metadata
Err:35 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/main amd64 c-n-f Metadata
  404  Not Found [IP: 188.174.173.78 443]
Ign:37 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe amd64 Packages
Ign:38 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe Translation-en
Ign:39 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/universe amd64 c-n-f Metadata
Ign:40 https://mirror.funkfreundelandshut.de/ubuntu focal-backports/multiverse amd64 c-n-f Metadata
Ign:41 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 Packages
Ign:41 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 Packages
Ign:42 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main Translation-en
Ign:42 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main Translation-en
Ign:43 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 c-n-f Metadata
Err:43 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 c-n-f Metadata
  404  Not Found [IP: 188.174.173.78 443]
Ign:44 https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted amd64 Packages
Ign:45 https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted Translation-en
Ign:46 https://mirror.funkfreundelandshut.de/ubuntu focal-security/restricted amd64 c-n-f Metadata
Ign:47 https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe amd64 Packages
Ign:48 https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe Translation-en
Ign:49 https://mirror.funkfreundelandshut.de/ubuntu focal-security/universe amd64 c-n-f Metadata
Ign:50 https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse amd64 Packages
Ign:51 https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse Translation-en
Ign:52 https://mirror.funkfreundelandshut.de/ubuntu focal-security/multiverse amd64 c-n-f Metadata
Ign:41 https://mirror.funkfreundelandshut.de/ubuntu focal-security/main amd64 Packages
Fetched 886 kB in 14s (63.8 kB/s)
Reading package lists...
E: Failed to fetch https://mirror.funkfreundelandshut.de/ubuntu/dists/focal/main/cnf/by-hash/SHA256/6cd27c7ff5478aaecc33cd98dc87e2e80a7a3f0211032d3480feda44d91f8aff  404  Not Found [IP: 188.174.173.78 443]
E: Failed to fetch https://mirror.funkfreundelandshut.de/ubuntu/dists/focal/restricted/binary-amd64/by-hash/SHA256/f641509ad1e4ab977debf3d852d1a7e47a02e9d941867f7b277d95cf37d64d49  404  Not Found [IP: 188.174.173.78 443]
E: Failed to fetch https://mirror.funkfreundelandshut.de/ubuntu/dists/focal/restricted/i18n/by-hash/SHA256/19afb5aa320e04ba441d1a9f[53](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:54)01b935bd53b6b58d5e145907d62e5ae4d0db4e  404  Not Found [IP: 188.174.173.78 443]
E: Failed to fetch https://mirror.funkfreundelandshut.de/ubuntu/dists/focal-updates/main/cnf/by-hash/SHA256/0c33e66ea0a49808f47430364534f06ae5af8a0b268cc273ba5848289f02d1a0  404  Not Found [IP: 188.174.173.78 443]
E: Failed to fetch https://mirror.funkfreundelandshut.de/ubuntu/dists/focal-backports/main/cnf/by-hash/SHA256/4c815edbfa0151b1a14e93e3e8c80dc4fce2de1ed5032d8a7138b[54](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:55)3218af17c  404  Not Found [IP: 188.174.173.78 443]
E: Failed to fetch https://mirror.funkfreundelandshut.de/ubuntu/dists/focal-backports/restricted/cnf/by-hash/SHA256/f332d98720671a387436468ca339789a285735974b12fe561011[55](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:56)ce2a[56](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:57)5adb  404  Not Found [IP: 188.174.173.78 443]
E: Failed to fetch https://mirror.funkfreundelandshut.de/ubuntu/dists/focal-security/main/cnf/by-hash/SHA256/e5453ad874e682b742ade35fc612aafc21d5dabbb[58](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:59)cdfd2ec7ee74bc014a256  404  Not Found [IP: 188.174.1[73](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:74).[78](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:79) 443]
E: Some index files failed to download. They have been ignored, or old ones used instead.
Error: Process completed with exit code [100](https://github.com/sippy/rtpproxy/actions/runs/3962692400/jobs/6789678728#step:3:101).
```

**Comment by till** [comment]
So it 404s somewhere?

**Comment by till** [comment]
@sobomax the logic is super simple, it checks if the URL is up:
https://github.com/lagged/apt-spy2/blob/master/lib/apt/spy2.rb#L88-L108

It does not actually check if the mirror carries a certain release, etc.. One thing is, this logic worked extremely well for many years at this point as it checks if a server is not just on a list but also up. ;)

You could probably make it query for the standard folders? Maybe with local output from `lsb_release -a`? Would you like to create a PR? :)

**Comment by sobomax** [comment]
> @sobomax the logic is super simple, it checks if the URL is up: https://github.com/lagged/apt-spy2/blob/master/lib/apt/spy2.rb#L88-L108
> 
> It does not actually check if the mirror carries a certain release, etc.. One thing is, this logic worked extremely well for many years at this point as it checks if a server is not just on a list but also up. ;)
> 
> You could probably make it query for the standard folders? Maybe with local output from `lsb_release -a`? Would you like to create a PR? :)

@till I might not be the best person to do this. First, I have zero knowledge about apt-get internals and zero time/interest to learn it. :( Just in case, my development platform of choice is FreeBSD since 2002. Last but not least I have never written a line of Ruby code yet.

The apt-spy2 came on my radar when I finally got annoyed by MicroSoft'ization of Ubuntu, making package installation the main source of random, but frequent failures of my projects on GH Actions. Searching for people's recommendations dealing with similar problems I came to apt-spy2, someone gave it a very good praise! And it looked like it helps (for a few days) until this happened. :( Which suggests that there is a more to it than just having server to respond to incoming HTTPS requests.

I can, however, give it another spin if someone extends "liveness" checking to be more comprehensive and report back any issues. I run 30-50 jobs daily, across 18.x, 20.x and 22.x, each doing apt update as it's first step, so it would give you very good coverage and uncover any remaining problems, if any.

Thanks for understanding and a constructive discussion!

**Comment by sobomax** [comment]
P.S. Here is another example of bizzare issues we are dealing with when using M$ apt-infrastructure:

```
Fetched 7894 kB in 2s (4940 kB/s)
Reading package lists...
E: Failed to fetch https://packages.microsoft.com/ubuntu/20.04/prod/dists/focal/main/binary-arm64/Packages.gz  Hash Sum mismatch
   Hashes of expected file:
    - Filesize:31455 [weak]
    - SHA[51](https://github.com/sippy/rtpproxy/actions/runs/4080347370/jobs/7032717985#step:3:52)2:9c9db0e775b00e4ec74e4f5d405da600ac851db2e13c862c7a8967cd1d34e6831378fd1d8349201f2960ab69067047493907b05d534be3a9e9e82324e9857346
    - SHA256:d27893aea19d202720dfb687c028ad6b7fd6966fa20c6f464701e42b219739d8
   Hashes of received file:
    - SHA512:a34bab67abd45842bb1ba63605cc1af8852835ce3b24439dbd95babe3fcba86201feb863176fb89333e80dbb41d2b444c59d9be41349055c1b704215631794cc
    - SHA256:0c7f62fa546391c8eb7dd5a00b2636b3da69d2acbc20d96fe49698f936571534
    - Filesize:31455 [weak]
   Last modification reported: Fri, 03 Feb 2023 01:48:13 +0000
   Release file created at: Fri, 03 Feb 2023 01:42:30 +0000
E: Failed to fetch https://packages.microsoft.com/ubuntu/20.04/prod/dists/focal/main/binary-all/Packages.gz  
E: Failed to fetch https://packages.microsoft.com/ubuntu/20.04/prod/dists/focal/main/binary-armhf/Packages.gz  
E: Some index files failed to download. They have been ignored, or old ones used instead.

**Comment by till** [comment]
Maybe you should run with the `--launchpad` flag.

If I remember correctly, the list of mirrors on there is a bit more stable and curated than the rrdns one (which always breaks). The rrdns list is why I built this tool 10 years ago.

---

Anyway, I stopped using Ubuntu myself, but I am willing to help regardless. All the code does is http requests, etc.. If you or anyone else would like to step up, lmk.

**Comment by parthshyara** [comment]
@till I took the liberty to create a PR for this. Please review!

**Comment by till** [comment]
> @till I took the liberty to create a PR for this. Please review!

Thanks very much. :) I reviewed, very minor changes. Let me know if this works for you!

**Comment by till** [comment]
@parthshyara Maybe I merged it too optimistically:
https://github.com/lagged/apt-spy2/actions/runs/4147973927/jobs/7175475454

Everything is reported as broken. ;)

**Comment by till** [comment]
I am gonna update my dev env to get ruby and take a looksie.

**Comment by till** [comment]
OK, I almost have it and will make a PR and let it marinate until tomorrow.

**Comment by till** [comment]
Alright, PR is up. This was too painful. I think I want to rewrite this tool in another language. 😂

@parthshyara If you wanna have a look, but no pressure. Thanks again for your initial PR!

**Comment by parthshyara** [comment]
@till I'm sorry that you had to go through this! I should've been more careful with the testing. Lesson learnt :)
Although, your PR is a silver lining as it fixes this and add other checks too! 😁

**Comment by parthshyara** [comment]
@till Thanks for fixing this! Please let us know once you release the version!

**Comment by till** [comment]
@parthshyara yes, will do!

I am fixing up some code etc.. I am also wondering if `--strict` should be a default.

**Comment by till** [comment]
OK, `0.8.0` was released.

Thanks again, @parthshyara! 🎉 

@sobomax If you want to take a look, you can now run with `--strict` to ensure the mirror carries the distribution you need. Let me know how it goes.

**Comment by till** [comment]
FYI: Just released `0.8.2` to fix an error I introduced by running `rubocop`.

**Comment by sobomax** [comment]
@till Thanks a bunch! Will totally do! :)
